src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -245,20 +245,40 @@ private void ReadDirectoryChangesCallback(uint errorCode, uint numBytes, AsyncRe
             }
         }
 
-        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer)
+        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer, uint numBytes)
         {
             Debug.Assert(buffer != null);
             Debug.Assert(buffer.Length > 0);
 
+            numBytes = Math.Min(numBytes, (uint)buffer.Length);
+
             fixed (byte* b = buffer)
             {
+                byte* pBufferLimit = b + numBytes;
+
                 Interop.Kernel32.FILE_NOTIFY_INFORMATION* info = (Interop.Kernel32.FILE_NOTIFY_INFORMATION*)b;
 
                 ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
 
                 // Parse each event from the buffer and notify appropriate delegates
                 do
                 {
+                    // Validate the data we received in case it's corrupted. This can happen
+                    // if we are watching files over the network with a poor connection.
+
+                    // Verify the info object is within the expected bounds
+                    if (info < b || (((byte*)info) + sizeof(Interop.Kernel32.FILE_NOTIFY_INFORMATION)) > pBufferLimit)
```

Comment:
I wonder if its worth moving `sizeof(Interop.Kernel32.FILE_NOTIFY_INFORMATION)` outside of the loop.

Reply:
`sizeof(Interop.Kernel32.FILE_NOTIFY_INFORMATION)` will JIT into a constant. You won't gain anything by manually hoisting it out.

Reply:
What's the reason that the compiler does not allow me to do this?
```
struct MyStruct { }

private unsafe void M()
{
    const int sizeOfMyStruct = sizeof(MyStruct);
    /// Do Something...
}
```

It gives me the following error:
> CS0133: The expression being assigned to 'sizeOfMyStruct' must be constant

Reply:
It is not C# constant, but it is a JIT-time constant.